Set Cilium as default CNI

embik opened the pull request:
**What this PR does / why we need it**:

This PR sets Cilium as the default CNI when creating a `Cluster` object and passing it through the webhook for defaulting.

**Which issue(s) this PR fixes**:
<!--optional, in `fixes #<issue number>` format, will close the issue(s) when PR gets merged-->
Fixes #

**What type of PR is this?**
/kind feature

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change? Then add your Release Note here**:
<!--
Write your release note. Release notes are being used to generate the changelog:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Set Cilium as default CNI for user clusters
```

**Documentation**:
<!--
Please do one of the following options:
- Add a link to the existing documentation
- Add a link to the kubermatic/docs pull request
- If no documentation change is applicable then add:
  - TBD (documentation will be added later)
  - NONE (no documentation needed for this PR)
-->
```documentation
NONE
```

embik committed:
Default cluster CNI to Cilium

Signed-off-by: Marvin Beckers <marvin@kubermatic.com>

embik committed:
Update unit tests for CNI mutation

Signed-off-by: Marvin Beckers <marvin@kubermatic.com>

wozniakjan committed:
conformance-tester: allow completed pods

Signed-off-by: Jan Wozniak <wozniak.jan@gmail.com>

wozniakjan commented on cmd/conformance-tester/pkg/runner/waiters.go:
```diff
@@ -122,7 +122,7 @@ func waitUntilAllPodsAreReady(ctx context.Context, log *zap.SugaredLogger, opts
 		for _, pod := range podList.Items {
 			// Ignore pods failing kubelet admission (KKP #6185)
-			if !util.PodIsReady(&pod) && !podFailedKubeletAdmissionDueToNodeAffinityPredicate(&pod, log) {
+			if !util.PodIsReady(&pod) && !podFailedKubeletAdmissionDueToNodeAffinityPredicate(&pod, log) && !util.PodIsCompleted(&pod) {
```
here the tests expected all user cluster pods to be in a state `Ready`, however, hubble adds a pod that ultimately succeeds and goes to `Completed` state. Added a small commit https://github.com/kubermatic/kubermatic/pull/12752/commits/33e8f4854fd25d5aa6cb5cd692992e26449a27ad permitting that.
```
{  context deadline exceeded; last error was: not all Pods are ready: [hubble-generate-certs-lcb7z]}
```
```
kube-system            hubble-generate-certs-l6x9h                  0/1     Completed   0             10m
```

wozniakjan committed:
cilium: set pod seccomp profile

Signed-off-by: Jan Wozniak <wozniak.jan@gmail.com>

wozniakjan commented on pkg/cni/cilium/cilium.go:
```diff
@@ -194,6 +194,11 @@ func GetAppInstallOverrideValues(cluster *kubermaticv1.Cluster, overwriteRegistr
 		},
+		"podSecurityContext": map[string]any{
+			"seccompProfile": map[string]any{
+				"type": "RuntimeDefault",
```
kkp tests expect pod seccomp profile to be set by default

https://github.com/kubermatic/kubermatic/blob/31f9faf5c8185733e59e2485dbd38180d4e2c51d/cmd/conformance-tester/pkg/tests/usercluster_controller.go#L149

> If we can, we should configure seccomp profiles in the default Helm values we pass. Not sure if the chart allows that.

[upstream chart seems to allows that](https://github.com/cilium/cilium/blob/a67489466c715854c4c472c939d32d35df3fe748/install/kubernetes/cilium/templates/cilium-agent/daemonset.yaml#L80C24-L80C42), followup to https://github.com/kubermatic/kubermatic/pull/8326

wozniakjan replied:
and it's actually needed on all cilium pods
https://github.com/cilium/cilium/commit/5aa1c236915444771e14ffb5fbe532e4e5de1c2d
```
{  function did not succeed after 3 attempts: expected seccomp profile on Pod kube-system/cilium-operator-85b8f66cd-hgkhq, got none
expected seccomp profile on Pod kube-system/hubble-generate-certs-g5pl8, got none
expected seccomp profile on Pod kube-system/hubble-relay-5ccf97b646-f6gnt, got none
expected seccomp profile on Pod kube-system/hubble-ui-869b75b895-trq2t, got none}
```

wozniakjan replied:
configured properly in https://github.com/kubermatic/kubermatic/pull/12752/commits/15cd1ef62ea4abbb849ea8693485fc65a0a90ff5

wozniakjan replied:
ok, `hubble-ui` and `hubble-generate-certs` doesn't have the option supported natively, with @embik we decided to skip these two as short-term remediation in 
https://github.com/kubermatic/kubermatic/pull/12752/commits/718db73794828f2602f057df65bbe8828888b997 and ideally we can try to contribute upstream.

wozniakjan committed:
cilium: chain CNI hostport plugin when kubeproxy still enabled

https://docs.cilium.io/en/latest/installation/cni-chaining-portmap/#portmap-hostport

Signed-off-by: Jan Wozniak <wozniak.jan@gmail.com>

wozniakjan commented on pkg/cni/cilium/cilium.go:
```diff
@@ -218,6 +244,8 @@ func GetAppInstallOverrideValues(cluster *kubermaticv1.Cluster, overwriteRegistr
 	} else {
 		values["kubeProxyReplacement"] = "disabled"
+		values["sessionAffinity"] = true
+		valuesCni["chainingMode"] = "portmap"
```
by default, KKP runs cilium with kube-proxy, this needs portmap plugin and session affinity enabled to pass conformance tests (otherwise, `HostPort` is not functional)
* https://docs.cilium.io/en/latest/installation/cni-chaining-portmap/#portmap-hostport
* https://github.com/cilium/cilium/issues/14287#issuecomment-1521757813

wozniakjan replied:
configured in https://github.com/kubermatic/kubermatic/pull/12752/commits/8ac82a3019a71e296d1e670fc9aef5b4adf47a7b

wozniakjan committed:
conformance-tester: skip seccomp for hubble-ui and hubble-generate-certs

Signed-off-by: Jan Wozniak <wozniak.jan@gmail.com>

cnvergence reviewed (approved):
/approve
